feat(ivy): allow the locale to be set via a global property

petebacondarwin opened the pull request:
The ivy compile time inlining tools work on the code after it has
been compiled and bundled (potentially minified and mangled).
It is only at this point in the processing that the current locale
can be specified.

// FW-1639

See https://github.com/angular/angular-cli/issues/15896

AndrewKushnir commented:
We discussed possible solutions with @mhevery and the proposal is to use `$localize.locale` as a way to communicate current locale to runtime code. The benefit of this approach is to have all localize-related logic grouped together under `$localize`. The processing of the `$localize.locale` might be different for compile-time inlining and runtime translations:

- for compile-time inlining, we can just replace `$localize.locale` with a string that represents current locale, so it would not be present in the final bundle, similar to `$localize` tokens
- for runtime translations, the `$localize` function would be available, so we can set the `locale` property to the current locale by adding additional code (in i18n inline logic or through CLI?)

The `_localeFactory` function (that is used to define the `LOCALE_ID` value) may access `$localize.locale` to get current locale (in the necessary branch of the `_localeFactory` function).

@petebacondarwin please let us know your opinion.

Thank you.

AndrewKushnir reviewed (changes requested):
See comment here: https://github.com/angular/angular/pull/33314#issuecomment-545083904. Thank you.

AndrewKushnir commented on packages/core/src/application_module.ts:
```diff
@@ -33,22 +35,30 @@ export function _keyValueDiffersFactory() {
+export function getGlobalLocale(): string {
+  return (ngI18nClosureMode && typeof goog !== 'undefined' && goog.LOCALE !== 'en') ?
+      goog.LOCALE :
+      ((ivyEnabled && $localize && $localize.locale) || DEFAULT_LOCALE_ID);
```
@petebacondarwin should we use `typeof $localize !== 'undefined'` to avoid an error in case `$localize` is not defined on a page?

petebacondarwin replied:
OK.

petebacondarwin commented on integration/cli-hello-world-ivy-i18n/angular.json:
```diff
@@ -4,6 +4,13 @@
+          "de": "src/locales/messages.de.json"
+        },
+        "sourceLocale": "en"
+      },
```
This is just preparatory - we are not yet using the built-in CLI inlining because it does not do the locale-inlining that this PR adds. @clydin - the CLI will need to be updated after this PR.

petebacondarwin commented on integration/cli-hello-world-ivy-i18n/angular.json:
```diff
@@ -72,7 +80,7 @@
               "vendorChunk": false,
               "buildOptimizer": true
-            }
+            } 
```
😱

petebacondarwin committed:
feat(ivy): allow the locale to be set via a global property

In the post-$localize world the current locale value is defined by setting
`$localize.locale` which is then read at runtime by Angular in the provider
for the `LOCALE_ID` token and also passed to the ivy machinery via`setLocaleId()`.

The $localize compile-time inlining tooling can replace occurrences of
`$localize.locale` with a string literal, similar to how translations
are inlined.

// FW-1639

See https://github.com/angular/angular-cli/issues/15896

petebacondarwin committed:
test: update ivy i18n integration test

The integration test now checks that the locale inlining is working.

petebacondarwin committed:
build: add non-ivy test script to package.json

This makes it easier to run non-ivy tests locally.

petebacondarwin committed:
refactor(ivy): i18n - create and use `isLocalize()` helper

petebacondarwin committed:
feat(ivy): i18n - inline current locale at compile-time

During compile-time translation inlining, the `$localize.locale`
expression will now be replaced with a string literal containing the
current locale of the translations.

mhevery commented on packages/core/src/application_module.ts:
```diff
@@ -33,22 +35,30 @@ export function _keyValueDiffersFactory() {
+ *   required, or just to provide their own `LOCALE_ID` provider.
+ */
+export function getGlobalLocale(): string {
+  return (ngI18nClosureMode && typeof goog !== 'undefined' && goog.LOCALE !== 'en') ?
```
Why are we making sure that `goog.LOCALE` is not `en`?

AndrewKushnir replied:
@mhevery there is a comment about that in function description. It used to be above that specific line, now it's a bit more subtle, so may be we should move it back (closer to that line of code).

petebacondarwin replied:
Will do

AndrewKushnir commented on integration/cli-hello-world-ivy-i18n/src/app/app.component.ts:
```diff
@@ -1,8 +1,10 @@
 @Component(
     {selector: 'app-root', templateUrl: './app.component.html', styleUrls: ['./app.component.css']})
 export class AppComponent {
+  constructor(@Inject(LOCALE_ID) public locale: string) {}
```
👍 thanks for adding this check!

AndrewKushnir reviewed (approved):
LGTM, thanks @petebacondarwin.

AndrewKushnir commented:
@petebacondarwin I've started g3 presubmits:

- [VE Presubmit](http://test/OCL:276375006:BASE:276375040:1571873282175:eae49e21)
- [Ivy Presubmit](http://test/OCL:276375006:BASE:276375389:1571873399538:df9c80d4)

But it looks like there is a couple failing tests (related to i18n) in `integration_test` job in CircleCI - could you please look into this? Please let me know if fixes would require new presubmit.

Thank you.

clydin commented on packages/core/src/application_module.ts:
```diff
@@ -33,22 +35,30 @@ export function _keyValueDiffersFactory() {
+export function getGlobalLocale(): string {
+  return (ngI18nClosureMode && typeof goog !== 'undefined' && goog.LOCALE !== 'en') ?
+      goog.LOCALE :
+      ((ivyEnabled && typeof $localize !== 'undefined' && $localize.locale) || DEFAULT_LOCALE_ID);
```
The undefined guard structure here is tightly coupled to the transformation in the inliner.  Alterations at a later date may unintentionally cause breakage.  I think a comment explaining that it needs to be kept in sync with the logic in the inliner may be a good idea.

IgorMinar replied:
+1

petebacondarwin replied:
Will do

IgorMinar commented on packages/localize/src/tools/test/translate/source_files/locale_plugin_spec.ts:
```diff
@@ -0,0 +1,86 @@
+    const output = transformSync(input, {plugins: [makeLocalePlugin('fr')]}) !;
+    expect(output.code).toEqual('const {\n  locale\n} = $localize;\nconst a = locale;');
+  });
+});
```
Very comprehensive set of tests! Thank you

petebacondarwin replied:
I added some more (and fixed a bug) related to the expression appearing on the LHS of an assignment.

There are a few alternatives of how to deal with this. Such assignments should only occur if the developer is doing runtime translation. We could strip out such assignments but I was nervous that this could break other code. Instead they are just being ignored. If the `$localize` symbol is not defined this would result in a runtime error - but anyone who has added this assignment should be aware of what they are doing. Also we are not officially supporting runtime for 9.0.0, so we can discuss and alter this later.

IgorMinar commented:
The integration tests are failing, plz take a look

petebacondarwin committed:
fixup! feat(ivy): i18n - inline current locale at compile-time

petebacondarwin committed:
fixup! feat(ivy): allow the locale to be set via a global property

petebacondarwin commented on integration/cli-hello-world-ivy-i18n/package.json:
```diff
@@ -50,8 +45,8 @@
-    "@angular/cli": "file:../../node_modules/@angular/cli",
+    "@angular-devkit/build-angular": "^0.900.0-next.14",
+    "@angular/cli": "^9.0.0-next.14",
     "@angular/compiler-cli": "file:../../dist/packages-dist/compiler-cli",
```
I'll create a new PR directly after this.

petebacondarwin replied:
@IgorMinar - as promised https://github.com/angular/angular/pull/33382

petebacondarwin committed:
fixup! test: update ivy i18n integration test

petebacondarwin commented:
Agh! The runtime translation setup added loads to the bundle sizes!
Cleaning that up...

filipesilva commented on integration/cli-hello-world-ivy-i18n/src/polyfills-runtime.ts:
```diff
@@ -0,0 +1,86 @@
+$localize.locale = 'fr';
+import {registerLocaleData} from '@angular/common';
+import localeFr from '@angular/common/locales/fr';
+registerLocaleData(localeFr);
```
This looks a lot like our previous CLI transform to add the locale:
https://github.com/angular/angular-cli/blob/99c174c03b101f0118dfb647313fc7a8d5233446/packages/ngtools/webpack/src/transformers/register_locale_data_spec.ts#L28-L44

I imagine it will be added in the same way. Probably in `main.ts` too, as the polyfills file is not guaranteed to exist.

petebacondarwin replied:
Yes, when we move over to using the CLI for all the translation stuff then I would expect this to be done by the CLI. But we need to tidy up the locale-id loose ends before we can do that.

I put the calls here because they had to be done before the bootstrap and I preferred to swap out this file than the `main.ts` in my configuration for runtime translation (since the `$localize.locale` is probably best done here).

petebacondarwin commented on integration/_payload-limits.json:
```diff
@@ -30,8 +30,8 @@
-        "main-es2015": 128258,
-        "polyfills-es2015": 42102
+        "main-es2015": 138032,
+        "polyfills-es2015": 37494
```
The polyfills.ts size is lower because I removed the runtime translations from it. The production build would inline them instead.

petebacondarwin commented on integration/_payload-limits.json:
```diff
@@ -30,8 +30,8 @@
         "runtime-es2015": 1485,
-        "main-es2015": 128258,
-        "polyfills-es2015": 42102
+        "main-es2015": 138032,
```
This file is bigger because we added to the application component's template: the date pipe and the locale id.

IgorMinar commented:
http://test/OCL:276502950:BASE:276502977:1571934258500:af4943d9
http://test/OCL:276502950:BASE:276503788:1571934589416:c35dd729

AndrewKushnir commented:
FYI, VE and Ivy presubmits are successful. Thank you.
